### PR TITLE
Add updateNodeRef for node link controller

### PR DESCRIPTION
**Pull request by enxebre**
Originally the logic for setting the nodeRef in a machine was split between the cluster-api node controller and opt-int external tooling that would set an annotation for the former controller to watch. There's no reason for this to keep leaving in two different places, furthermore the cluster-api node controller is deprecated. This PR unifies the current logic and behaviour in one single nodelink controller.
Counterpart PR https://github.com/openshift/cluster-api/pull/47

Note:
This PR is to unify and simplify current code in one single nodelink controller while maintaining current behaviour to simplify farther development and maintainability.
Once we promote MHC, we could decide we don't care about the machine controller taking care of recreating a machine if something out of band deletes the cloud instance, and remove the newNodeReady != currentNodeReady check, and just watch machines instead of nodes and exclusively rely on MHC

### CODE REVIEW

File: pkg/controller/nodelink/nodelink_controller.go [hunk 1]
```diff
@@ -234,10 +246,43 @@ func (r *ReconcileNodeLink) Reconcile(request reconcile.Request) (reconcile.Resu
 		}
 	}
 
-	// TODO(alberto): bring nodeRef logic here and drop cluster-api node controller
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileNodeLink) updateNodeRef(machine *mapiv1beta1.Machine, node *corev1.Node) error {
```
Comment (bison) [inline pkg/controller/nodelink/nodelink_controller.go]: Could we add a comment here, or at the call site, about the reason for "touching" the machine resource on node status change. It's not immediately clear.

Comment (enxebre) [reply pkg/controller/nodelink/nodelink_controller.go]: done https://github.com/openshift/machine-api-operator/pull/336/files#diff-69885f1e6e1f6b4cb889ce80021c95a1R273

File: pkg/controller/nodelink/nodelink_controller.go [hunk 2]
```diff
@@ -474,3 +519,16 @@ func (r *ReconcileNodeLink) listMachinesByField(key, value string) ([]mapiv1beta
 	}
 	return machineList.Items, nil
 }
+
+// isNodeReady returns true if a node is ready; false otherwise.
+func isNodeReady(node *corev1.Node) bool {
+	if node == nil || &node.Status == nil {
```
[the remark this answers is not in the document]
Comment (enxebre) [reply pkg/controller/nodelink/nodelink_controller.go]: removed, don't really need the check

File: pkg/controller/nodelink/nodelink_controller.go [hunk 3]
```diff
@@ -210,6 +213,15 @@ func (r *ReconcileNodeLink) Reconcile(request reconcile.Request) (reconcile.Resu
 		return reconcile.Result{}, nil
 	}
 
+	if err := r.updateNodeRef(machine, node); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error updating nodeRef for machine %q and node %q: %v", machine.GetName(), node.GetName(), err)
+	}
+
+	if !node.DeletionTimestamp.IsZero() {
```
Comment (vikaschoudhary16) [inline pkg/controller/nodelink/nodelink_controller.go]: Why updating NodeRef on the machine but skipping to update machine annotation on node?

Comment (vikaschoudhary16) [reply pkg/controller/nodelink/nodelink_controller.go]: nevermind. Now i see handling of deletion timsestamp case in `updateNodeRef`.

File: pkg/controller/nodelink/nodelink_controller_test.go [hunk 4]
```diff
@@ -705,3 +706,67 @@ func TestIndexMachineByInternalIP(t *testing.T) {
 		}
 	}
 }
+
+func TestUpdateNodeRef(t *testing.T) {
+	testCases := []struct {
+		machine            *mapiv1beta1.Machine
+		node               *corev1.Node
+		nodeRef            *corev1.ObjectReference
+		nodeReadinessCache map[string]bool
```
Comment (vikaschoudhary16) [inline pkg/controller/nodelink/nodelink_controller_test.go]: where is this being used?

Comment (enxebre) [reply pkg/controller/nodelink/nodelink_controller_test.go]: https://github.com/openshift/machine-api-operator/pull/336/files#diff-69885f1e6e1f6b4cb889ce80021c95a1R273

**Pull request merged**